src/main/java/org/folio/circulation/services/AllowedServicePointsService.java

```diff
@@ -40,167 +46,198 @@ public class AllowedServicePointsService {
+      ? itemRepository.fetchById(request.getItemId())
+        .thenApply(r -> r.map(List::of))
+      : itemFinder.getItemsByInstanceId(UUID.fromString(request.getInstanceId()), true);
+  }
+
+  private CompletableFuture<Result<Map<RequestType, Set<AllowedServicePoint>>>>
+  extractAllowedServicePoints(RequestPolicy requestPolicy) {
+
+    log.debug("extractAllowedServicePoints:: parameters requestPolicy: {}", requestPolicy);
 
     List<RequestType> allowedTypes = Arrays.stream(RequestType.values())
       .filter(requestPolicy::allowsType)
       .toList();
+
     if (allowedTypes.isEmpty()) {
       log.info("fetchAllowedServicePoints:: allowedTypes is empty");
-
       return ofAsync(new EnumMap<>(RequestType.class));
     }
-    log.info("fetchAllowedServicePoints:: allowedTypes={}", allowedTypes);
 
-    return fetchPickupLocationServicePointsIds()
-      .thenApply(r -> r.map(servicePointIds -> mapToAllowedServicePoints(allowedTypes,
-        servicePointIds)));
+    var allowedServicePointsInPolicy = requestPolicy.getAllowedServicePoints();
+    if (allowedServicePointsInPolicy.isEmpty()) {
+      log.info("extractAllowedServicePoints:: allowedServicePointsInPolicy is empty");
+
+      return fetchAllowedServicePoints()
+        .thenApply(r -> r.map(allowedServicePoints -> groupAllowedServicePointsByRequestType(
+          allowedTypes, allowedServicePoints)));
+    }
+
+    Set<String> allowedServicePointsIds = allowedServicePointsInPolicy.values().stream()
+      .flatMap(Collection::stream)
+      .collect(Collectors.toSet());
+
+    return fetchPickupLocationServicePointsByIds(allowedServicePointsIds)
+      .thenApply(r -> r.map(servicePoints -> groupAllowedServicePointsByRequestType(
+        allowedTypes, servicePoints, allowedServicePointsInPolicy)));
   }
 
-  private Map<RequestType, Set<String>> mapToAllowedServicePoints(List<RequestType> allowedTypes,
-    Set<String> servicePointIds) {
+  private Map<RequestType, Set<AllowedServicePoint>> groupAllowedServicePointsByRequestType(
+    List<RequestType> allowedTypes, Set<AllowedServicePoint> allowedServicePoints) {
 
-    log.debug("mapToAllowedServicePoints:: parameters: allowedTypes={}, servicePointsIds={}",
-      () -> asJson(allowedTypes), () -> asJson(servicePointIds));
+    log.debug("groupAllowedServicePointsByRequestType:: parameters allowedTypes: {}, " +
+        "servicePointsIds: {}", () -> asJson(allowedTypes), () -> asJson(allowedServicePoints));
 
-    Map<RequestType, Set<String>> allowedServicePoints = new EnumMap<>(RequestType.class);
-    allowedTypes.forEach(requestType -> allowedServicePoints.put(requestType, servicePointIds));
-    log.info("fetchAllowedServicePoints:: result: {}", () -> asJson(allowedServicePoints));
+    Map<RequestType, Set<AllowedServicePoint>> groupedAllowedServicePoints =
+      new EnumMap<>(RequestType.class);
+    allowedTypes.stream()
+      .filter(requestType -> !allowedServicePoints.isEmpty())
```

Comment:
This condition will have same value for all elements of the stream, so we can check it outside of the stream:

```
    if (allowedServicePoints.isEmpty()) {
      return new EnumMap<>(RequestType.class);
    }

    return allowedTypes.stream()
      .collect(Collectors.toMap(identity(), type -> allowedServicePoints));
```

Reply:
done